### ISSUE TITLE
11

### ISSUE THREAD

**Issue by SteveHong1901** [SteveHong1901/STAT0023-Coursework-1#5]
import numpy as np

class MulticlassLogisticRegression:
    def __init__(self, learning_rate=0.001, num_iterations=1000):
        self.learning_rate = learning_rate
        self.num_iterations = num_iterations

    def _softmax(self, z):
        exps = np.exp(z - np.max(z))  # for numerical stability
        return exps / np.sum(exps, axis=1, keepdims=True)

    def _cross_entropy(self, y_true, y_prob):
        m = y_true.shape[0]
        return -np.sum(y_true * np.log(y_prob)) / m

    def fit(self, X, y):
        m, n = X.shape
        self.classes = np.unique(y)
        k = len(self.classes)  # Number of classes

        # One-hot encode y
        y_encoded = np.zeros((m, k))
        for i, val in enumerate(y):
            y_encoded[i, val] = 1

        # Add bias term to X
        X_bias = np.insert(X, 0, 1, axis=1)

        # Initialize weights
        self.weights = np.random.randn(n + 1, k)

        for i in range(self.num_iterations):
            logits = X_bias.dot(self.weights)
            probabilities = self._softmax(logits)

            gradient = X_bias.T.dot(probabilities - y_encoded) / m

            self.weights -= self.learning_rate * gradient

            if i % 100 == 0:
                loss = self._cross_entropy(y_encoded, probabilities)
                print(f"Iteration {i}: Loss {loss:.6f}")

    def predict(self, X):
        X_bias = np.insert(X, 0, 1, axis=1)
        logits = X_bias.dot(self.weights)
        probabilities = self._softmax(logits)
        return np.argmax(probabilities, axis=1)

# Usage example
# X_train, y_train = ...  # Your training data
# X_test = ...           # Your testing data
# model = MulticlassLogisticRegression()
# model.fit(X_train, y_train)
# predictions = model.predict(X_test)
